processing-service-core/api/models/tenant.js

```diff
@@ -0,0 +1,351 @@
+}
+
+/**
+ * Delete tenant specific configs for the provided tenantId.
+ * Deletes all tenant specific configuration for tenant that has been provided.
+ *
+ * @param {String} tenantId String The ID of the tenant.
+ * @returns {*|d.promise|Function|promise|a|h} a promise that will be resolved or rejected based on result of deletion.
+ **/
+function deleteTenantConfigs(tenantId) {
+    var tenantConfig = Q.defer();
+    var deletionPromises = [];
+    tenantConfigStoreModel.getTenantConfigs(tenantId)
+            .then(function (tenantConfigs) {
+                var tenantConfigsMap = createMap(tenantConfigs);
+                for (var key in tenantConfigsMap) {
+                    deletionPromises.push(tenantConfigStoreModel.deleteTenantConfig(tenantId, key));
+                }
+                logger.debug("Successfully deleted tenant specific config for tenant " + tenantId);
+                tenantConfig.resolve();
+            })
+            .all(deletionPromises)
+            .fail(function (errorResponse) {
+                logger.debug("Unable to remove tenant specific config for tenant {}", tenantId);
+                tenantConfig.reject(errorResponse);
+            })
+            .done();
+    return tenantConfig.promise;
+}
+
+/**
+ * Retrieve effective tenant specific config.
+ * Returns the effective tenant specific config for this key. That can be either a custom value specified for the tenant or the default.
+ *
+ * @param {String} tenantId String The ID tenant.
+ * @param {String} key String The custom configuration key.
+ * @returns {*|d.promise|Function|promise|a|h} a promise that will be resolved or rejected based on result of retrieval.
+ **/
+function getEffectiveTenantConfig(tenantId, key) {
+    var tenantConfig = Q.defer();
+    var effectiveConfig = {};
+    logger.debug("Retrieving effective tenant config for tenant " + tenantId + " with key " + key);
+    tenantConfigStoreModel.getTenantConfig(tenantId, key)
+            .then(function (tenantConfigResult) {
+                effectiveConfig.value = tenantConfigResult;
+                effectiveConfig.valueType = "CUSTOM";
+                tenantConfig.resolve(effectiveConfig);
+            })
+            .fail(function (errorResponse) {
+                logger.debug("No tenant specific config for tenant " + tenantId + " with key " + key + " was found, returning default"
+                        + "value");
+                if (errorResponse instanceof ApiError && errorResponse.type === apiErrorTypes.ITEM_NOT_FOUND) {
+                    globalConfigStoreModel.getGlobalConfig(key)
+                            .then(function (tenantDefaultConfig) {
+                                for (var key in tenantDefaultConfig) {
+                                    effectiveConfig.value = tenantDefaultConfig[key];
+                                    effectiveConfig.valueType = "DEFAULT";
+                                    tenantConfig.resolve(effectiveConfig);
```

Comment:
This works because there's just one property expected to be on 'tenantDefaultConfig' I think but that resolve should be outside the for statement in case this changes in the future and because it won't make sense to ever call resolve twice.